src/lib.rs

```diff
@@ -294,17 +344,48 @@ impl<T> Sender<T> {
 
 impl<T> Drop for Sender<T> {
     fn drop(&mut self) {
-        // SAFETY: The reference won't be used after the channel is freed in this method
+        // SAFETY: The receiver only ever frees the channel if we are in the MESSAGE or
+        // DISCONNECTED states. If we are in the MESSAGE state, then we called
+        // mem::forget(self), so we should not be in this function call. If we are in the
+        // DISCONNECTED state, then the receiver either received a MESSAGE so this statement is
+        // unreachable, or was dropped and observed that our side was still alive, and thus didn't
+        // free the channel.
         let channel = unsafe { self.channel_ptr.as_ref() };
 
         // Set the channel state to disconnected and read what state the receiver was in
-        match channel.state.swap(DISCONNECTED, SeqCst) {
+        // ORDERING: we don't need release ordering here since there are no modifications we
+        // need to make visible to other thread, and the Err(RECEIVING) branch handles
+        // synchronization independent of this cmpxchg
+        //
+        // EMPTY ^ 001 = DISCONNECTED
+        // RECEIVING ^ 001 = UNPARKING
+        // DISCONNECTED ^ 001 = EMPTY (invalid), but this state is never observed
+        match channel.state.fetch_xor(0b001, Relaxed) {
             // The receiver has not started waiting, nor is it dropped.
             EMPTY => (),
             // The receiver is waiting. Wake it up so it can detect that the channel disconnected.
-            RECEIVING => unsafe { channel.take_waker() }.unpark(),
+            RECEIVING => {
+                // See comments in Sender::send
+
+                fence(Acquire);
+
+                let waker = unsafe { channel.take_waker() };
+
+                // We still need release ordering here to make sure our read of the waker happens
+                // before this, and acquire ordering to ensure the unparking of the receiver
+                // happens after this.
+                channel.state.swap(DISCONNECTED, AcqRel);
+
+                // `unpark` has an implicit release ordering, so the store of the DISCONNECTED
+                // state happens-before we unpark the receiver
```

Comment:
This comment is now outdated. We don't rely on this implicit undocumented release ordering that is supposedly not part of the stable API of `unpark`.